### PR TITLE
implement Enum.empty? as Enum.count == 0

### PR DESCRIPTION
**Pull request by xbinxu**

### CODE REVIEW

**Comment by joshnuss** [comment]
@xbinxu `count/1` traverses the whole list. Using `reduce` with `:halt`, is faster becaus only the first item is looked up.

**Comment by xbinxu** [comment]
@joshnuss, reduce with :halt seems not work for Map, I've test it for a huge map with 100k items, it takes a long time.

```elixir
 m100k = (1..100000) |> Enum.map(fn(x) -> {x, x + 1} end) |> Enum.into(%{})

  :timer.tc(fn -> Enum.empty?(m100k) end)
  {3162, false}
```

but Enum.count is stable as i expected.

```elixir
:timer.tc(fn -> Enum.count(m100k) end)
  {14, 100000}
```

File: lib/elixir/lib/enum.ex [hunk 1]
```diff
@@ -633,9 +633,14 @@ defmodule Enum do
...
   def empty?(enumerable) do
-    Enumerable.reduce(enumerable, {:cont, true},
-      fn(_, _) -> {:halt, false} end)
-    |> elem(1)
+    case Enumerable.count(enumerable) do
+      {:ok, value} when is_integer(value) ->
+        value == 0
+      {:error, module} ->
+        Enumerable.reduce(enumerable, {:cont, true},
+          fn(_, _) -> {:halt, false} end)
```
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: Please call `module.reduce` here. :)

**Comment by josevalim** [comment]
@xbinxu thank you! I have added one tiny comment and we are good to go!

**Comment by ericmj** [comment]
I think this change is wrong since `Enumerable.count` runs in linear time and the previous version of this function always ran in constant time.

(Sorry if I'm missing some context, I've been on vacation.)

**Comment by josevalim** [comment]
Enumerable.count is not linear. It is meant to be implemented only if not
linear. Otherwise you should return the error tuple.

>

-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by ericmj** [comment]
Are you sure? Documentation says it runs in linear time and things like
streams implement it which can only run in linear.
>
>

-- 
Eric Meadows-Jönsson

**Comment by josevalim** [comment]
Enum.count is linear because that's the only guarantee we can give.
Enumerable.count is not. :)

On Thursday, November 12, 2015, Eric Meadows-Jönsson <
notifications@github.com> wrote:

> Are you sure? Documentation says it runs in linear time and things like
> streams implement it which can only run in linear.
> >
> >
>
> --
> Eric Meadows-Jönsson
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elixir-lang/elixir/pull/3948#issuecomment-156040357>.
>


-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by ericmj** [comment]
We should make this clear in documentation. I can't find anything in docs
that reflect this.



-- 
Eric Meadows-Jönsson

**Comment by fishcakez** [comment]
@ericmj the relevant doc is here http://elixir-lang.org/docs/stable/elixir/Enumerable.html#count/1 and asks nicely but could be more detailed.  Current method is O(n) for maps: https://groups.google.com/d/msg/elixir-lang-talk/icehVNnDmEw/cYEDly4IFQAJ

**Pull request merged**